filters/include/pcl/filters/impl/spherical_voxel_grid.hpp

```diff
@@ -0,0 +1,349 @@
+          continue;
+
+      PointT shiftedPoint = input_->points[*it];
+
+      shiftedPoint.getVector4fMap () -= filter_origin_;
+
+      // Convert to spherical coordinates
+      float r = shiftedPoint.getVector3fMap ().norm ();
+
+      float theta = std::acos (shiftedPoint.z / r);
+
+      // Convert range of atan2 to ( 0, 2PI ]
+      float phi = std::atan2 (shiftedPoint.y, shiftedPoint.x) + M_PI;
+
+      // Get index
+      int rIdx = static_cast<int> (std::floor (r / leaf_size_r_));
+      int thetaIdx = static_cast<int> (std::floor (theta / leaf_size_theta_));
+      int phiIdx = static_cast<int> (std::floor (phi / leaf_size_phi_));
+
+      int idx = ((thetaIdx * leaf_phi_divisions_) + phiIdx) + (leaf_theta_divisions_ * leaf_phi_divisions_ * rIdx);
+
+      index_vector.push_back (pcl::detail::cloud_point_index_idx (idx, *it));
+    }
+  }
+
+  // Second pass: sort the index_vector vector using value representing target cell as index
+  // in effect all points belonging to the same output cell will be next to each other
+  std::sort (index_vector.begin (), index_vector.end (), std::less<pcl::detail::cloud_point_index_idx> ());
+
+  // Third pass: count output cells
+  // we need to skip all the same, adjacenent idx values
+  unsigned int total = 0;
+  unsigned int index = 0;
+  // first_and_last_indices_vector[i] represents the index in index_vector of the first point in
+  // index_vector belonging to the voxel which corresponds to the i-th output point,
+  // and of the first point not belonging to.
+  std::vector<std::pair<unsigned int, unsigned int> > first_and_last_indices_vector;
+  // Worst case size
+  first_and_last_indices_vector.reserve (index_vector.size ());
+  while (index < index_vector.size ())
+  {
+    unsigned int i = index + 1;
+    while (i < index_vector.size () && index_vector[i].idx == index_vector[index].idx)
+      ++i;
+    if (i - index >= min_points_per_voxel_)
+    {
+      ++total;
+      first_and_last_indices_vector.push_back (std::pair<unsigned int, unsigned int> (index, i));
+    }
+    index = i;
+  }
+
+  // Fourth pass: compute centroids, insert them into their final position
+  output.points.resize (total);
+
+  index = 0;
+  for (unsigned int cp = 0; cp < first_and_last_indices_vector.size (); ++cp)
+  {
+    // calculate centroid - sum values from all input points, that have the same idx value in index_vector array
+    unsigned int first_index = first_and_last_indices_vector[cp].first;
+    unsigned int last_index = first_and_last_indices_vector[cp].second;
+
+    //Limit downsampling to coords
+    if (!downsample_all_data_)
+    {
+      Eigen::Vector4f centroid (Eigen::Vector4f::Zero ());
+
+      for (unsigned int li = first_index; li < last_index; ++li)
+        centroid += input_->points[index_vector[li].cloud_point_index].getVector4fMap ();
+
+      centroid /= static_cast<float> (last_index - first_index);
+      output.points[index].getVector4fMap () = centroid;
+    }
```

Comment:
There's a `pcl::compute3DCentroid` to compute this for you. If you already have a std::vector with all the indices it's a single function call.

Reply:
But we don't have such a vector, or..?

Reply:
This remark was more highlighting potential benefits of already having the point indices for each voxel available by this stage. Basically trying to provided compelling arguments for an implementation change in that direction.